google/cloud/spanner/client.cc

```diff
@@ -128,6 +128,37 @@ std::shared_ptr<Connection> MakeConnection(
                                                     std::move(stub));
 }
 
+StatusOr<CommitResult> RunTransaction(
+    Client client, Transaction::ReadWriteOptions const& opts,
+    std::function<TransactionAction(Client, Transaction)> const& f) {
+  Transaction txn = MakeReadWriteTransaction(opts);
+  TransactionAction action;
+#if GOOGLE_CLOUD_CPP_HAVE_EXCEPTIONS
+  try {
+#endif
+    action = f(client, txn);
+#if GOOGLE_CLOUD_CPP_HAVE_EXCEPTIONS
+  } catch (...) {
+    client.Rollback(txn);  // ignore error
+    throw;
+  }
+#endif
+  switch (action.action) {
+    case TransactionAction::kCommit: {
+      // TODO(#357): Automatically retry if Commit() aborts. Note: it is
+      // not a good idea to simply cap the number of retries. Instead, it
+      // is better to limit the total amount of wall time spent retrying.
+      return client.Commit(txn, action.mutations);
+    }
+    case TransactionAction::kRollback: {
+      auto status = client.Rollback(txn);
```

Comment:
Rollbacks can also fail and may need to be retried.  They are easier to retry, of course.

Reply:
Are you referring to "RPC-level" retry that we should apply to all remote calls, or something different?

I don't know that `Rollback()` can ever return a server-initiated "aborted" like `Commit()`, which is what we're dealing with in this particular retry loop.  Indeed, I'm hard pressed to think of any way a server-initiated `Rollback()` error could be retried.  The transaction has probably already been rolled back.

Reply:
Yes, I was referring to the RPC-level retry.  A `Rollback` that returns `StatusCode::kUnavailable` is a (I think): not an operation that is known to have completed successfully (or with a permanent error), and it can be retried safely.

Reply:
Acknowledged.